thunder/transforms/prune_prologue_checks.py

```diff
@@ -0,0 +1,56 @@
+import thunder
+from thunder.core import prims as prims
+
+
+class PrunePrologueChecks(thunder.core.transform_common.Transform):
+    """A Transform to prune Prologue checks
+
+    This transform removes prologue checks and can be applied when the user
+    controls the environment enough to ensure that these checks would always
+    succeed. By default, we remove all checks that use the module.
+    """
+
+    def __init__(self, prune_all_checks=False):
+        self.prune_all_checks = prune_all_checks
+
+    def transform_traces_pre_prologue(self, prologue_trace, compute_trace, epilogue_trace, **kwargs):
+        def is_check(bsym):
+            return bsym.sym in {
+                prims.check_tensor_shape_and_metadata,
+                prims.check_string_value,
+                prims.check_number_type_and_value,
+                prims.check_len,
+                prims.check_literal_like,
+            }
+
+        if not self.prune_all_checks:
+            bsyms_to_skip = set()
+            module_member_names = set()
+
+            for bsym in prologue_trace.bound_symbols:
+                if bsym.sym in {prims.unpack_trivial, prims.unpack_cache_info}:
+                    # These don't have inputs but need to be skipped to not trigger false positives
+                    continue
+                elif all(i.name in module_member_names for i in bsym.flat_proxy_args):
+                    # This has the special case of no proxy inputs, which is the case for unpack_function_obj,
+                    # the root of module_member_names
```

Comment:
Do you think we should add an `assert bsym.sym == unpack_function_obj`?

Reply:
all not empty, but yes, fair.